File: apps/api/services/connectors/registry.py
"""Adapter registry: source_id -> adapter instance. No connector-specific logic here."""

from typing import Protocol

from .contract import ConnectorInput, ConnectorOutput


class ConnectorAdapter(Protocol):
    """Adapter interface: fetch(input) -> output. Optional test_connection for admin."""

    async def fetch(self, input: ConnectorInput) -> ConnectorOutput:
        ...

    def test_connection(self, config: dict) -> tuple[bool, str | None]:
        """Return (success, error_message). Used by admin before save."""
        ...


_adapters: dict[str, ConnectorAdapter] = {}


def register(source_id: str, adapter: ConnectorAdapter) -> None:
    _adapters[source_id] = adapter


def get_adapter(source_id: str) -> ConnectorAdapter | None:
    return _adapters.get(source_id)


def register_adapters() -> None:
    """Register all adapters (call at app startup)."""
    from services.connectors.mock_bitrix import MockBitrixAdapter
    from services.connectors.google_drive import GoogleDriveAdapter

    register("bitrix", MockBitrixAdapter())
    register("google_drive", GoogleDriveAdapter())

File: apps/web/lib/server/connectors/registry.ts
/**
 * Adapter registry: source_id -> adapter. No connector-specific logic. Server-only.
 */

import type { ConnectorAdapter } from "./contract";
import { MockBitrixAdapter } from "./mock-bitrix";
import { GoogleDriveAdapter } from "./google-drive";

const adapters = new Map<string, ConnectorAdapter>();

export function register(sourceId: string, adapter: ConnectorAdapter): void {
  adapters.set(sourceId, adapter);
}

export function getAdapter(sourceId: string): ConnectorAdapter | null {
  if (!registered) registerAdapters();
  return adapters.get(sourceId) ?? null;
}

let registered = false;

export function registerAdapters(): void {
  if (registered) return;
  registered = true;
  register("bitrix", new MockBitrixAdapter());
  const driveAdapter = new GoogleDriveAdapter();
  register("google_drive", driveAdapter);
  register("google_sheets", driveAdapter);
}
